# 3D Structure Prediction and Disorder Analysis Toolkit

This toolkit provides two primary functionalities: 
- 3D Structure Prediction Using Multiple Methods: UI tool which allows to collect predictions from up to 3 models (ESMfold, OmegaFold, Swiss-Model)
- Disorder Prediction and Visualization: This functionality runs IUPred for a provided UniProt ID or PDB file and colors the structure by disorder using PyMOL.


## 3D Structure Prediction Using Multiple Methods

### Requirements:

#### Swiss-Model (https://swissmodel.expasy.org/)
To use this service, you'll need to register and obtain an API token. Follow the steps below to get started:
- Create account at `https://swissmodel.expasy.org/`
- Go to `https://swissmodel.expasy.org/account`
- Copy your API token 
- Save it as SWISSMODEL_EXPASY_API_TOKEN in your environment

#### ESM Metagenomic (ESMFold) (https://github.com/facebookresearch/esm)
ESMFold is a state-of-the-art method for predicting protein structures from metagenomic data. It leverages deep learning techniques and the ESM model to accurately predict protein structure.

Fortunately, no additional steps are required.

#### OmegaFold (https://github.com/HeliXonProtein/OmegaFold?tab=readme-ov-file)

To use this model you need follow the instructions below:

```
git clone https://github.com/HeliXonProtein/OmegaFold
cd OmegaFold
python setup.py install
```
This command will download the weight from https://helixon.s3.amazonaws.com/release1.pt to 
```~/.cache/omegafold_ckpt/model.pt``` and load the model.

#### Packages

Make sure to create a conda environment with the command:
```
conda env create -f environment.yml
```

### Usage
To start our application, run the following command:
```
python 3d_structure_analysis.py
```

## Disorder Prediction and Visualization

### Overview
The second component of this toolkit allows users to analyze the intrinsic disorder
of protein structures. It utilizes the IUPred tool to predict disorder regions in protein sequence. The resulting 
predictions are then wizualized by coloring the protein structure in PyMOL.

### Requirements
- biopython
- requests
- pymol

### Example
```bash
# For a UniProt ID
python disorder_analysis/disorder_analysis.py P12345

# For a PDB file
python disorder_analysis/disorder_analysis.py path/to/your.pdb
```

## Folder architecture
```
ðŸ“¦ 
â”œâ”€Â 3d_structure_analysis.py (main script with UI)
â”œâ”€Â LICENSE
â”œâ”€Â README.md
â”œâ”€Â disorder_analysis 
â”‚Â Â â”œâ”€Â disorder_analysis.py (iupred2a runner, can be used separately)
â”‚Â Â â””â”€Â iupred2a
â”‚Â Â Â Â Â â”œâ”€Â .idea
â”‚Â Â Â Â Â â”‚Â Â â”œâ”€Â inspectionProfiles
â”‚Â Â Â Â Â â”‚Â Â â”‚Â Â â””â”€Â profiles_settings.xml
â”‚Â Â Â Â Â â”‚Â Â â””â”€Â iupred2a.iml
â”‚Â Â Â Â Â â”œâ”€Â LICENSE
â”‚Â Â Â Â Â â”œâ”€Â P53_HUMAN.seq
â”‚Â Â Â Â Â â”œâ”€Â README
â”‚Â Â Â Â Â â”œâ”€Â __init__.py
â”‚Â Â Â Â Â â”œâ”€Â data
â”‚Â Â Â Â Â â”‚Â Â â”œâ”€Â anchor2_energy_matrix
â”‚Â Â Â Â Â â”‚Â Â â”œâ”€Â anchor2_interface_comp
â”‚Â Â Â Â Â â”‚Â Â â”œâ”€Â iupred2_long_energy_matrix
â”‚Â Â Â Â Â â”‚Â Â â”œâ”€Â iupred2_short_energy_matrix
â”‚Â Â Â Â Â â”‚Â Â â”œâ”€Â long_histogram
â”‚Â Â Â Â Â â”‚Â Â â””â”€Â short_histogram
â”‚Â Â Â Â Â â”œâ”€Â iupred2a.py
â”‚Â Â Â Â Â â””â”€Â iupred2a_lib.py
â”œâ”€Â environment.yml
â”œâ”€Â structure_prediction (scripts to run models/sends API requests, can be used separately)
â”‚Â Â â”œâ”€Â esm.py
â”‚Â Â â”œâ”€Â omega_fold.py
â”‚Â Â â””â”€Â swissmodel.py
â””â”€Â utils.py (functions used by 3d_structure_analysis.py)
```
Â©generated by [Project Tree Generator](https://woochanleee.github.io/project-tree-generator)
